### PR TITLE
Initial support for automated testing in MyLA with Cypress

### PR DESCRIPTION
**Pull request by mcrwong**
This is the final update for getting cypress testing for MyLA. 
This includes snapshot tests for the grade distribution, resources accessed, and new assignment planning, whose snapshots should be updated whenever the related svgs are changed. This also tests some standard login functionality, with the user: donald07 and password: root

These tests are based upon the data from the myla_test_data_2020_03_11.sql file.

Overall cypress isn't terribly different from selenium, except with a fancier gui and a lower barrier to entry.

Cypress still hasn't fixed the bug with running all tests that breaks the cypress-plugin-snapshot, so when running tests, click the individual test file rather than the all tests button. To run cypress, both [cypress](https://www.cypress.io/) and the [cypress-plugin-snapshot](https://github.com/meinaart/cypress-plugin-snapshots) need to be installed. The cypress client can be opened with the npm run cypress:open command.

### CODE REVIEW

**Comment by justin0022** [comment]
Hi @mcrwong, I'm looking forward to reviewing this, but unfortunately won't be able to get to it till next week.

**Comment by jonespm** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cypress/plugins/index.js  1
- cypress/integration/myla/basic_test.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ITS_Teaching_And_Learning/my-learning-analytics/pullRequest?prid=5411197&bid=17740823)

**Comment by pushyamig** [comment]
@mcrwong I am trying to run this locally for reviewing it. These are the steps I did for the set up
1. installed Cypress and cypress-plugin-snapshot
  "cypress": "4.4.1",
"cypress-plugin-snapshots": "1.3.0",
2. installed the same DB dump that was committed as part of this PR, I have user `donald07` and root PW is `root`
3. opened cypress client using `npm run cypress:open` I can see the client is launched. 
Screen Shot 2020-04-28 at 4 21 33 PM
4. click on `basic_test.js` leads to opening a blank browser window and eventually Timeout error
Screen Shot 2020-04-28 at 4 24 06 PM

Did I miss anything that I am not able to run Cypress test?

**Comment by mcrwong** [comment]
> @mcrwong I am trying to run this locally for reviewing it. These are the steps I did for the set up
> 
> 1. installed Cypress and cypress-plugin-snapshot
>    "cypress": "4.4.1",
>    "cypress-plugin-snapshots": "1.3.0",
> 2. installed the same DB dump that was committed as part of this PR, I have user `donald07` and root PW is `root`
> 3. opened cypress client using `npm run cypress:open` I can see the client is launched.
>    Screen Shot 2020-04-28 at 4 21 33 PM
> 4. click on `basic_test.js` leads to opening a blank browser window and eventually Timeout error
>    Screen Shot 2020-04-28 at 4 24 06 PM
> 
> Did I miss anything that I am not able to run Cypress test?

Try switching to the electron browser instead of chrome for now. There should be a button in the top right of the cypress client to do so.

I'll look into it more, but you seem to have followed the same steps as me.

**Comment by pushyamig** [comment]
@mcrwong That worked! I just need to add `donald07` user from django console and create a password `root` and after that test successfully ran
Screen Shot 2020-04-28 at 4 46 31 PM

**Comment by pushyamig** [comment]
> Can you also try installing cypress outside the repo and running the default tests that come with the default cypress installation? That should tell whether the issue is my code or something else.

I followed a [tutorial](https://www.valentinog.com/blog/cypress/) on cypress and I was able to run that without any issue this morning on Chrome browser

**Comment by pushyamig** [comment]
OK, Think I figured out what is the reason for the problem with CHrome Browser when running cypress with Myla. I think Initially I was just trying to run the Cypress from Tutorial project and I used Chrome( Default) with that. The default port was 5000 and I never stopped the npm run from that.

later I started to test the Myla cypress setup with Chrome again which was running on 5001. Shouldn't not conflict

Somehow I need to stop the former(Tutorial) cypress run to make it work. So I again tested cypress Myla run with chrome and it worked.

**Comment by pushyamig** [comment]
I think the cypress and cypress-plugin-snapshot should be part of package.json, why did you choose that option?

**Comment by mcrwong** [comment]
> I think the cypress and cypress-plugin-snapshot should be part of package.json, why did you choose that option?

If I remember correctly, when I added those to my package.json, the docker build would fail.

**Comment by pushyamig** [comment]
>If I remember correctly, when I added those to my package.json, the docker build would fail.
Hmm.. Need to understand why is that.

**Comment by pushyamig** [comment]
I tried adding the cypress related lib to the `package.json` and it builds fine and App starts up fine. But When I try to open cypress client from docker container (`docker exec -it webpack_watcher npm run cypress:open`) then it fails to say that it needs some additional dependency

```
> dashboard@0.0.1 cypress:open /usr/src/app
> cypress open

It looks like this is your first time using Cypress: 4.5.0

  ✖  Verifying Cypress can run /root/.cache/Cypress/4.5.0/Cypress
    → Cypress Version: 4.5.0
Your system is missing the dependency: Xvfb

Install Xvfb and run Cypress again.

Read our documentation on dependencies for more information:

https://on.cypress.io/required-dependencies

If you are using Docker, we provide containers with all required dependencies installed.
```
To me, this is a different Question altogether, We can run it locally for testing it without getting into the docker route. But the ultimate goal for this setup is to eliminate the manual testing route the BSA has to go through ( by adding more test case to Cypress). 

Adding the cypress library when not using in docker would make the build time more. Document the Cypress testing the Readme.

**Comment by pushyamig** [comment]
@mcrwong with selenium tests you did for the Mwrite project, how was your script used for testing in QA? Were they run manually in local machine or there was a server setup done for testing in QA?

File: cypress/plugins/index.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+/**
+ * @type {Cypress.PluginConfig}
+ */
+ const { initPlugin } = require('cypress-plugin-snapshots/plugin');
+
+module.exports = (on, config) => {
+  initPlugin(on, config);
+  return config;
+};
```
Comment (pushyamig) [inline cypress/plugins/index.js:23]: Can you remove the semi-colons as it is not needed? and fix the indentation as it will show as errors with [Standard](https://standardjs.com/)library but work fine with or with out it

Comment (justin0022) [reply cypress/plugins/index.js:23]: `standard --fix` should auto-fix the formatting issues @mcrwong

File: cypress.json [hunk 2]
```diff
@@ -0,0 +1,39 @@
+{
+	"baseUrl": "http://localhost:5001",
+	"ignoreTestFiles": [
+	"**/__snapshots__/**",
+	"**/__image_snapshots__/**"
+	],
+	"env": {
```
Comment (pushyamig) [inline cypress.json:7]: Looks to me like this is default copy/paste from [Github](https://github.com/meinaart/cypress-plugin-snapshots). These are an option if we want to configure anything.

**Comment by pushyamig** [comment]
@mcrwong  Nice Work! You work is a good start for enabling others to write more test. 
There are some Question need to answer 
1. how we want to make the E2E testing working from respective Myla instances (dev/qa) for automated testing
2. If we want to test with different user/URL is there a way we can make it a configurable option.
3. Support of Docker etc.

**Comment by pushyamig** [comment]
@mcrwong Please add running cypress test in README and any helpful documentation they can refer to

File: README.md [hunk 3]
```diff
@@ -162,6 +162,12 @@ After about 30-60 seconds the crons should have completed and you should have da
 3. The url for configuring copyright info must be `/copyright/` since that is used in the `base.html` for pulling the info. [Read more here](https://simpleisbetterthancomplex.com/tutorial/2016/10/04/how-to-use-django-flatpages-app.html)
 
 
+## Accessibility
+
+### Keyboard Navigation
+Users that use Safari or Firefox on macOS that would like to enable keyboard navigation in MyLA will have to follow the steps [here](https://github.com/tl-its-umich-edu/my-learning-analytics/wiki/Accessibility).
+
```
Comment (pushyamig) [inline README.md:169]: I am not sure why this change is as part of your PR

**Comment by jonespm** [comment]
Are the files in __snapshots__ and __image_snapshots__ regenerated on each run? If so I'd remove __snapshots__/basic_test.js.snap which is part of this commit and add these paths to the .gitignore.

**Comment by pushyamig** [comment]
> There are some Question need to answer 
>1. how we want to make the E2E testing working from respective Myla instances (dev/qa) for >automated testing
>2. If we want to test with different user/URL is there a way we can make it a configurable option.
>3. Support of Docker etc. 

@mcrwong answers need not be from you. Your input is valuable

**Comment by mcrwong** [comment]
> Are the files in **snapshots** and **image_snapshots** regenerated on each run? If so I'd remove **snapshots**/basic_test.js.snap which is part of this commit and add these paths to the .gitignore.

I believe that the snapshots are only regenerated if the test fails and the tester manually chooses to switch to the new snapshots. It will also generate snapshots on the first run as well. I thought to ensure that everyone is testing against the same snapshots, that the snapshots should be uploaded.

File: README.md (hunk 3 above)
Comment (mcrwong) [reply README.md:169]: Might be a remnant from when I was squashing commits together.

**Comment by mcrwong** [comment]
> @mcrwong with selenium tests you did for the Mwrite project, how was your script used for testing in QA? Were they run manually in local machine or there was a server setup done for testing in QA?

@pushyamig 
They were run locally, but it wouldn't take much changing to switch selenium to run chrome in headless mode and set up the CI to just run the script.

**Comment by pushyamig** [comment]
@mcrwong your latest commit is showing a lot more change that are not associated with your Changes.
